src/main.c

```diff
@@ -247,10 +365,145 @@ void WriteInterfaces()
     free(InterfaceArray);
 }
 
+void CreateVmSwitchPacketFragment(PEVENT_RECORD ev, unsigned long LowerIfIndex)
+{
+    wchar_t Buffer[8192];
+    PROPERTY_DATA_DESCRIPTOR Desc;
+    int Err;
+    PNDIS_NET_BUFFER_LIST_8021Q_INFO pNblVlanInfo;
+
+    // Get VLAN from OOB
+    unsigned long OobLength;
+    Desc.PropertyName = (unsigned long long)L"OOBDataSize";
+    Desc.ArrayIndex = ULONG_MAX;
+    Err = TdhGetProperty(ev, 0, NULL, 1, &Desc, sizeof(OobLength), (PBYTE)&OobLength);
+    if (Err != NO_ERROR) {
+        printf("TdhGetProperty OobLength failed with %u\n", Err);
+        return;
+    }
+
+    if (OobLength > sizeof(OobData)) {
+        printf("OOB data of %lu bytes too large to fit in hardcoded buffer of size %lu\n", OobLength, (unsigned long)sizeof(OobData));
+        return;
+    }
+
+    Desc.PropertyName = (unsigned long long)L"OOBData";
+    Desc.ArrayIndex = ULONG_MAX;
+    Err = TdhGetProperty(ev, 0, NULL, 1, &Desc, OobLength, (PBYTE)&OobData);
+    if (Err != NO_ERROR) {
+        printf("TdhGetProperty Metadata failed with %u\n", Err);
+        return;
+    }
+
+    pNblVlanInfo = (PNDIS_NET_BUFFER_LIST_8021Q_INFO)&OobData[Ieee8021QNetBufferListInfo];
+    VMSwitchPacketFragment.VmSwitchVlanId = pNblVlanInfo->TagHeader.VlanId;
+
+    // SourcePortId 
+    Desc.PropertyName = (unsigned long long)L"SourcePortId";
+    Desc.ArrayIndex = ULONG_MAX;
+    Err = TdhGetProperty(ev, 0, NULL, 1, &Desc, sizeof(VMSwitchPacketFragment.VmSwitchSourcePortId), (PBYTE)&VMSwitchPacketFragment.VmSwitchSourcePortId);
+    if (Err != NO_ERROR) {
+        printf("TdhGetProperty VmSwitchSourcePortId failed with %u\n", Err);
+        return;
+    }
+
+    // DestinationCount
+    Desc.PropertyName = (unsigned long long)L"DestinationCount";
+    Desc.ArrayIndex = ULONG_MAX;
+    Err = TdhGetProperty(ev, 0, NULL, 1, &Desc, sizeof(VMSwitchPacketFragment.VmSwitchDestinationCount), (PBYTE)&VMSwitchPacketFragment.VmSwitchDestinationCount);
+    if (Err != NO_ERROR) {
+        printf("TdhGetProperty VmSwitchDestinationCounts failed with %u\n", Err);
+        return;
+    }
+
+    // Parse packet information to be written into the interface (only if a matching interface
+    // doesn't exist, i.e. if one is about to be created. Otherwise this packet info is not
+    // needed).
+    if (!Pass2) {
+        struct INTERFACE* Iface = GetInterface(LowerIfIndex);
+        if (Iface == NULL) {
+
+            // SourceNicName
```

Comment:
Good optimization. But why don't we do this one-time parsing in AddInterface? The purpose of the parsing would be more obvious there, it would automatically only happen when we need it, and we would avoid the extra call of GetInterface.